### PR TITLE
RUN-1227: use `unique_leaky_ptr` for all `PowerModeVoter`s

### PR DESCRIPTION
**Pull request by Domiii**
* https://linear.app/replay/issue/RUN-1227/divergence-frameschedulerimpl-destroys-powermodevoter

### CODE REVIEW

**Comment by Domiii** [comment]
@bhackett1024 This is a first sample of the idea. I don't think its as good as it could be since I wanted to inherit from `unique_ptr` instead, but was unable to get that to work. What do you think?

**Review by kannanvijayan — COMMENTED** [review]
Looks good, just a couple nits:

1. Following with our naming convention, calling this `record_replay_unique_leaky_pointer` or `recordreplay::UniqueLeakyPtr` might be appropriate.
2. Places where this is assigned is likely to have a `unique_ptr` on the RHS, so we probably need an explicit assignment overload that takes a `unique_ptr&&` and moves it.
3. Probably need an `explicit` constructor taking a `T*`, and an assignment overload for the same.

File: base/record_replay.h [hunk 1]
```diff
@@ -203,6 +204,52 @@ class SCOPED_LOCKABLE AutoUnlockMaybeEventsDisallowed {
...
+  // Return the stored pointer.
+  T* operator->() const noexcept { return get(); }
+
+  /// Return the stored pointer.
+  T* get() const noexcept { return p.get(); }
+
+  // dtor
+  ~unique_leaky_ptr() {
+    // -> leak the allocated memory, before destructing `unique_ptr`.
+    p.release();
```
Comment (bhackett1024) [inline base/record_replay.h:249]: This should only leak the pointer when AreEventsDisallowed, in keeping with the existing places where we leak these unique ptrs.  This keeps the non-recording case from being impacted, reduces the impact when recording/replaying, and allows us to turn on/off this leaking behavior via the record/replay features so we can see any correctness / performance impacts from behavior changes like this.  The comment on the class also needs to be updated.

**Review by bhackett1024 — CHANGES_REQUESTED** [review]
The class name looks good to me, and I think that having a unique_ptr member is clearer than inheriting from unique_ptr.

File: base/record_replay.h (hunk 1 above)
Comment (Domiii) [reply base/record_replay.h]: Done

Comment (Domiii) [reply base/record_replay.h]: Feel free to re-review

**Review by bhackett1024 — CHANGES_REQUESTED** [review]
Can you also remove the code where we manually check for AreEventsDisallowed in the destructors for some of the classes which have these voters?

btw with github if you want someone to re-review a patch you need to add them as a reviewer again, otherwise it won't show up in the review queue.

**Review by bhackett1024 — CHANGES_REQUESTED** [review]
Can you address the comment from my last review?

**Comment by Domiii** [comment]
@bhackett1024 
> Can you address the comment from my last review?

Sure. Here were your previous two comments:

### "This should only leak the pointer when `AreEventsDisallowed`"

I did that, did I not?

But now you are saying:

> Can you [...] remove the code where we manually check for `AreEventsDisallowed` in the destructors for some of the classes which have these voters?

What are "some of the classes"? Or are you saying, this should be like a flag (maybe in form of a type parameter), or...?


### "The comment on the class also needs to be updated."

I did that, too. But maybe you can be more specific about what you want updated? :shrug:

**Review by bhackett1024 — APPROVED** [review]

**Pull request merged**